contracts/lbp/LBPManager.sol

```diff
@@ -0,0 +1,290 @@
+            )
+        );
+
+        lbp.updateWeightsGradually(
+            _startTimeEndTime[0],
+            _startTimeEndTime[1],
+            _endWeights
+        );
+
+        return address(lbp);
+    }
+
+    /**
+     * @dev                             Subtracts the primeDaoFee and adds liquidity to the LBP.
+     * @param _tokenList                Numerically sorted array (ascending) containing two addresses:
+                                            - The address of the project token being distributed.
+                                            - The address of the funding token being exchanged for the project token.
+     * @param _projectTokenIndex        Index for the _tokenList array for the funding token.
+     * @param _sender                   Address of the liquidity provider.
+     * @param _userData                 UserData object that specifies the type of join.
+     */
+    function addLiquidity(
+        IERC20[] memory _tokenList,
+        uint8 _projectTokenIndex,
+        address _sender,
+        bytes memory _userData
+    ) external onlyAdmin {
+        require(!poolFunded, "LBPManager: pool has already been funded");
+        poolFunded = true;
+
+        IVault vault = lbp.getVault();
+
+        if (primeDaoFeePercentage != 0) {
+            // Transfer primeDaoFee to beneficiary.
+            _tokenList[_projectTokenIndex].transferFrom(
+                _sender,
+                beneficiary,
+                _feeAmountRequired(_projectTokenIndex)
+            );
+        }
+
+        for (uint8 i; i < _tokenList.length; i++) {
+            _tokenList[i].transferFrom(_sender, address(this), amounts[i]);
+            _tokenList[i].approve(address(vault), amounts[i]);
+        }
+
+        IVault.JoinPoolRequest memory request = IVault.JoinPoolRequest({
+            maxAmountsIn: amounts,
+            userData: _userData,
+            fromInternalBalance: false, // It is not possible to add liquidity through the internal Vault balance.
+            assets: _tokenList
+        });
+
+        vault.joinPool(lbp.getPoolId(), address(this), address(this), request);
+    }
+
+    /**
+     * @dev                             exit pool or remove liquidity from pool
+     * @param _tokenList                Numerically sorted array (ascending) containing two addresses:
+                                            - The address of the project token being distributed.
+                                            - The address of the funding token being exchanged for the project token.
+     * @param _receiver                 Address of the liquidity receiver, after exiting the LBP.
+     * @param _toInternalBalance        Balancer's Vault option to send the liquidity from the LBP to the _receivers internal Vault balance.
+     * @param _userData                 UserData object that specifies the type of exit.
+     */
+    function removeLiquidity(
+        IERC20[] memory _tokenList,
```

Comment:
Hmm, do we need to check tokenList here, what if we remove Liquidity not from all tokens? Or if we try to removeLiquidity for tokens that were not previously sent?

Reply:
P.S. as I understand we can call this function few times?

Reply:
Balancer will return an error if we try to withdraw with tokens not used before. Also, I think it's not possible to only withdraw one half of the pools liquidity. I do wonder now why we pass the `tokenList` every time and not just take them from the state variables defined by the call `initializeLBP`? That would make it cleaner as well right?

Reply:
@codebuster22 maybe you have a reason why we shouldn't just use the state variables?

Reply:
And we can only call this function once because of the require statement that checks if the manager has a balance of pool tokens.

```
        require(
            lbp.balanceOf(address(this)) > 0,
            "LBPManager: manager does not have any pool tokens to remove liquidity"
        );
```

Reply:
Yep, I think it will be a lot more clear without passing the list of tokens...
Maybe it is done because of the inheritance reasons, but I don't think that it is needed

Reply:
We can call this function only once if we pass the whole list of tokens, but what if we have 4 tokens in the pool and we pass only 2?

Reply:
Balancer LBP can't have 4 tokens in a pool. So in case of the LBP this would not work. Also I think a LBP with multiple tokens doesn't really make sense

Reply:
Could you explain what you mean by inheritance reason? That the state variable would be available in inherited contracts so you might want to pass them as arguments?

Reply:
Hmm if it can have only 2 tokens, why don't we limit array sizes? I will add an issue to do it

Reply:
LBP will only have two tokens. Not more than that. The reason I added _tokenList was, to make testing easier. Once we have all the functionality implemented. We were going to optimize it.
Here we are using `Exact BPT/Pool tokens in for Amounts out`. So we specify how much `BPT Tokens` we want to give, and the minimum amount of `Tokens` we want out. As this PR was already heavy. We wanted to complete the functionality and try to do housekeeping work and optimization in next.